core/src/main/java/com/scalar/db/common/TransactionExecutor.java

```diff
@@ -1,20 +1,35 @@
 package com.scalar.db.common;
 
+import com.google.common.util.concurrent.Uninterruptibles;
 import com.scalar.db.api.CrudOperable;
 import com.scalar.db.api.DistributedTransaction;
 import com.scalar.db.api.DistributedTransactionManager;
+import com.scalar.db.exception.transaction.CommitConflictException;
+import com.scalar.db.exception.transaction.CrudConflictException;
 import com.scalar.db.exception.transaction.RollbackException;
 import com.scalar.db.exception.transaction.TransactionException;
+import com.scalar.db.exception.transaction.UnknownTransactionStatusException;
 import com.scalar.db.transaction.singlecrudoperation.SingleCrudOperationTransactionUtils;
 import com.scalar.db.util.ThrowableConsumer;
 import com.scalar.db.util.ThrowableFunction;
+import java.util.concurrent.TimeUnit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public final class TransactionExecutor {
 
   private static final Logger logger = LoggerFactory.getLogger(TransactionExecutor.class);
 
+  private static final int DEFAULT_RETRY_INITIAL_INTERVAL_MILLIS = 100;
+  private static final int DEFAULT_RETRY_MAX_INTERVAL_MILLIS = 1000;
+  private static final int DEFAULT_RETRY_MULTIPLIER = 2;
+  private static final int DEFAULT_RETRY_MAX_RETRIES = 5;
+
+  private static int retryInitialIntervalMillis = DEFAULT_RETRY_INITIAL_INTERVAL_MILLIS;
+  private static int retryMaxIntervalMillis = DEFAULT_RETRY_MAX_INTERVAL_MILLIS;
+  private static int retryMultiplier = DEFAULT_RETRY_MULTIPLIER;
+  private static int retryMaxRetries = DEFAULT_RETRY_MAX_RETRIES;
```

Comment:
Mutable class variables are potentially not safe, I think. These parameters are like global variables and could be unexpectedly changed by other code.

I guess there are 2 types of use case of executeWithRetries():
1. Use the default parameters
2. Use custom parameters

For 1, this PR has `executeWithRetries(DistributedTransactionManager, ThrowableFunction<CrudOperable<?>, T, TransactionException>)`. I think it's okay to directly use the constraints (e.g., `DEFAULT_RETRY_INITIAL_INTERVAL_MILLIS`), instead of using the mutable class variables. If custom parameter is needed for retry, consider the use case 2.

For 2, this PR also has `executeWithRetries(DistributedTransactionManager, ThrowableFunction<CrudOperable<?>, T, TransactionException>, int, int, int, int)`, which doesn't use the class variables.

With changing `executeWithRetries(DistributedTransactionManager, ThrowableFunction<CrudOperable<?>, T, TransactionException>)` to directly use the default value constraints, probably we can remove the mutable class variables. What do you think?

Reply:
> I guess there are 2 types of use case of executeWithRetries():
> 1. Use the default parameters
> 2. Use custom parameters

I was considering the third use case:
3. Change the default parameters and use them.

I thought that if you wanted to use custom parameters but always use the same ones, you would need to call `executeWithRetries(DistributedTransactionManager, ThrowableFunction<CrudOperable<?>, T, TransactionException>, int, int, int, int)` with the same parameters every time, which would be a bit redundant. Therefore, for better usability, I added setters to allow changing the default values.

So, this is a discussion about safety vs usability, right?

Of course, one idea is to make `TransactionExecutor` a stateful class (by turning the parameters into instance fields), but I’d prefer not to do that because it’s not very handy.

Reply:
> one idea is to make `TransactionExecutor` a stateful class (by turning the parameters into instance fields), but I’d prefer not to do that because it’s not very handy.

I think this is straightforward. For usability, providing a default static instance is a common way and would improve the usability?

```java
public final class TransactionExecutor {
  private static final int DEFAULT_RETRY_INITIAL_INTERVAL_MILLIS = 100;
  private static final int DEFAULT_RETRY_MAX_INTERVAL_MILLIS = 1000;
  private static final int DEFAULT_RETRY_MULTIPLIER = 2;
  private static final int DEFAULT_RETRY_MAX_RETRIES = 5;

  private final int retryInitialIntervalMillis;
  private final int retryMaxIntervalMillis;
  private final int retryMultiplier;
  private final int retryMaxRetries;

  // Or `INSTANCE` is also common?
  public static final TransactionExecutor DEFAULT = new TransactionExecutor(
        DEFAULT_RETRY_INITIAL_INTERVAL_MILLIS,
        DEFAULT_RETRY_MAX_INTERVAL_MILLIS,
        DEFAULT_RETRY_MULTIPLIER,
        DEFAULT_RETRY_MAX_RETRIES);

  public TransactionExecutor(int retryInitialIntervalMillis, int retryMaxIntervalMillis,
      int retryMultiplier, int retryMaxRetries) {
    this.retryInitialIntervalMillis = retryInitialIntervalMillis;
    this.retryMaxIntervalMillis = retryMaxIntervalMillis;
    this.retryMultiplier = retryMultiplier;
    this.retryMaxRetries = retryMaxRetries;
  }

  ...
```

With this, `DEFAULT.executeWithRetries(...)` is used for normal execution, and custom TransactionExecutor instances with custom parameters can be created, shared and used for special cases.

> I thought that if you wanted to use custom parameters but always use the same ones

BTW, I still don't get in which case to globally use customized default parameters. Probably I'm missing something, but can you elaborate on the use case?

Reply:
> BTW, I still don't get in which case to globally use customized default parameters. Probably I'm missing something, but can you elaborate on the use case?

@komamitsu I imagine a use case where we want to use the same retry parameters across an application but need to change the max retry count from the default 5 to 10.

That said, while I can see the potential for this use case, I’m actually going to use `TransactionExecutor` with the default parameters in another module. Therefore, we don’t need to make the default parameters configurable for now. Additionally, since this class is primarily intended for internal use (as mentioned in the additional notes), I’ll remove the mutable class variables and the setters. Thank you for discussing this!